components/style/properties/gecko.mako.rs

```diff
@@ -663,6 +665,121 @@ fn static_assert() {
+            self.gecko.mBorderImageOutset.data_at_mut(${side.index})
+                .copy_from(&other.gecko.mBorderImageOutset.data_at(${side.index}));
+        % endfor
+    }
+
+    pub fn set_border_image_repeat(&mut self, v: longhands::border_image_repeat::computed_value::T) {
+        use properties::longhands::border_image_repeat::computed_value::RepeatKeyword;
+        use gecko_bindings::structs::{NS_STYLE_BORDER_IMAGE_REPEAT_STRETCH, NS_STYLE_BORDER_IMAGE_REPEAT_REPEAT};
+        use gecko_bindings::structs::{NS_STYLE_BORDER_IMAGE_REPEAT_ROUND, NS_STYLE_BORDER_IMAGE_REPEAT_SPACE};
+        let mut keywords = vec![];
+
+            % for side in range(2):
+                let k = match v.${side} {
+                    RepeatKeyword::Stretch => NS_STYLE_BORDER_IMAGE_REPEAT_STRETCH,
+                    RepeatKeyword::Repeat => NS_STYLE_BORDER_IMAGE_REPEAT_REPEAT,
+                    RepeatKeyword::Round => NS_STYLE_BORDER_IMAGE_REPEAT_ROUND,
+                    RepeatKeyword::Space => NS_STYLE_BORDER_IMAGE_REPEAT_SPACE,
+                };
+                keywords.push(k);
+            % endfor
+
+            self.gecko.mBorderImageRepeatH = keywords[0] as u8;
+            self.gecko.mBorderImageRepeatV = keywords[1] as u8;
+    }
+
+    pub fn copy_border_image_repeat_from(&mut self, other: &Self) {
+        self.gecko.mBorderImageRepeatH = other.gecko.mBorderImageRepeatH;
+        self.gecko.mBorderImageRepeatV = other.gecko.mBorderImageRepeatV;
+    }
+
+    pub fn set_border_image_width(&mut self, v: longhands::border_image_width::computed_value::T) {
+        use properties::longhands::border_image_width::computed_value::SingleComputedValue;
+
+        % for side in SIDES:
+        match v.${side.index} {
+            SingleComputedValue::Auto =>
+                self.gecko.mBorderImageWidth.data_at_mut(${side.index}).set_value(CoordDataValue::Auto),
+            SingleComputedValue::LengthOrPercentage(l) =>
+                l.to_gecko_style_coord(&mut self.gecko.mBorderImageWidth.data_at_mut(${side.index})),
+            SingleComputedValue::Number(n) =>
+                self.gecko.mBorderImageWidth.data_at_mut(${side.index}).set_value(CoordDataValue::Factor(n)),
+        }
+        % endfor
+    }
+
+    pub fn copy_border_image_width_from(&mut self, other: &Self) {
+        % for side in SIDES:
+            self.gecko.mBorderImageWidth.data_at_mut(${side.index})
+                .copy_from(&other.gecko.mBorderImageWidth.data_at(${side.index}));
+        % endfor
+    }
+
+    pub fn set_border_image_slice(&mut self, v: longhands::border_image_slice::computed_value::T) {
+        use gecko_bindings::structs::{NS_STYLE_BORDER_IMAGE_SLICE_NOFILL, NS_STYLE_BORDER_IMAGE_SLICE_FILL};
+        use properties::longhands::border_image_slice::computed_value::PercentageOrNumber;
+
+        for (i, corner) in v.corners.iter().enumerate() {
+            match *corner {
+                PercentageOrNumber::Percentage(p) =>
```

Comment:
Oh the `LengthOrNumber` is actually single value of `border_image_outset`. It's not actually stored in `style/values/..`, it's stored in `border_image_outset` longhand. Should I implement it still?